[#301] iRODS 5: Document upgrade_irods.py script (main) (#353)
irods/irods_docs: docs/getting_started/upgrading.md

#

Upgrading is handled by the host Operating System via the package manager.  Depending on your package manager, your configuration files will have been preserved with your local changes since the last installation.  Please see [Changing the zone_key and negotiation_key](installation.md#changing-the-zone_key-and-negotiation_key) for information on server-server authentication.

All servers in a Zone must be running the same version of iRODS.  Using inconsistent versions within a Zone may work, but is not tested.  First, upgrade the iRODS Catalog Service Provider, then upgrade all the iRODS Catalog Service Consumers.

It is best practice to stop an iRODS server before upgrading as it will allow the graceful completion of any ongoing transfers or requests.

> Upgrades coming from the APT and YUM repositories will cause the server to shut down. This is intentional as the administrator is required to apply catalog schema updates before restarting the server. A catalog schema update could be a relatively heavy operation and will require an amount of time on large installations (hundreds of millions of records) that should be handled within a declared maintenance window.

Reviewer: ```suggestion
Upgrades of packages from APT and YUM repositories will stop the server as part of the package upgrade process. This is intentional as the administrator is required to apply any catalog schema updates before restarting the server. A catalog schema update could be a relatively heavy operation and will require an amount of time on large installations (hundreds of millions of records) that should be handled within a declared maintenance window.
```

## Upgrading to iRODS 5 and later

> iRODS 5 does not upgrade the catalog or `server_config.json` on server startup. This functionality was deliberately separated to make the upgrade process more deterministic and manageable.

Reviewer: ```suggestion
iRODS 5 does not upgrade the catalog or `server_config.json` on server startup. This functionality has been kept separate deliberately to make the upgrade process more deterministic and manageable.
```

Upgrading only requires two steps:

1. Install the new packages.
1. As the service account user, run `python3 scripts/upgrade_irods.py`.

The `upgrade_irods.py` script is idempotent.

!!! Warning

>     The iRODS server will not start if it detects a difference in the catalog schema version.

Reviewer: ```suggestion
    The iRODS server will not start if it detects an unexpected catalog schema version.
```

Author: Done.
